File: src/bindings/software_scale_context.cc
#include "software_scale_context.h"
#include "frame.h"
#include "common.h"


namespace ffmpeg {

Napi::FunctionReference SoftwareScaleContext::constructor;

// === Init ===

Napi::Object SoftwareScaleContext::Init(Napi::Env env, Napi::Object exports) {
  Napi::Function func = DefineClass(env, "SoftwareScaleContext", {
    // Lifecycle
    InstanceMethod<&SoftwareScaleContext::AllocContext>("allocContext"),
    InstanceMethod<&SoftwareScaleContext::GetContext>("getContext"),
    InstanceMethod<&SoftwareScaleContext::InitContext>("initContext"),
    InstanceMethod<&SoftwareScaleContext::FreeContext>("freeContext"),

    // Operations
    InstanceMethod<&SoftwareScaleContext::Scale>("scale"),
    InstanceMethod<&SoftwareScaleContext::ScaleFrame>("scaleFrame"),

    // Utility
    InstanceMethod(Napi::Symbol::WellKnown(env, "dispose"), &SoftwareScaleContext::Dispose),
  });
  
  constructor = Napi::Persistent(func);
  constructor.SuppressDestruct();
  
  exports.Set("SoftwareScaleContext", func);
  return exports;
}

// === Lifecycle ===

SoftwareScaleContext::SoftwareScaleContext(const Napi::CallbackInfo& info) 
  : Napi::ObjectWrap<SoftwareScaleContext>(info) {
  // Constructor does nothing - user must explicitly call allocContext() or getContext()
}

SoftwareScaleContext::~SoftwareScaleContext() {
  // Manual cleanup if not already done
  if (ctx_ && !is_freed_) {
    sws_freeContext(ctx_);
    ctx_ = nullptr;
  }
  // RAII handles cleanup
}

// === Methods ===

Napi::Value SoftwareScaleContext::AllocContext(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  // Free existing context if any
  if (ctx_ && !is_freed_) { sws_freeContext(ctx_); ctx_ = nullptr; is_freed_ = true; };
  
  SwsContext* new_ctx = sws_alloc_context();
  if (!new_ctx) {
    Napi::Error::New(env, "Failed to allocate SwsContext").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (ctx_ && !is_freed_) { sws_freeContext(ctx_); } ctx_ = new_ctx; is_freed_ = false;
  
  return env.Undefined();
}

Napi::Value SoftwareScaleContext::GetContext(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 7) {
    Napi::TypeError::New(env, "Expected 7 arguments (srcW, srcH, srcFormat, dstW, dstH, dstFormat, flags)")
        .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  int srcW = info[0].As<Napi::Number>().Int32Value();
  int srcH = info[1].As<Napi::Number>().Int32Value();
  int srcFormat = info[2].As<Napi::Number>().Int32Value();
  int dstW = info[3].As<Napi::Number>().Int32Value();
  int dstH = info[4].As<Napi::Number>().Int32Value();
  int dstFormat = info[5].As<Napi::Number>().Int32Value();
  int flags = info[6].As<Napi::Number>().Int32Value();
  
  // Free existing context if any
  if (ctx_ && !is_freed_) { sws_freeContext(ctx_); ctx_ = nullptr; is_freed_ = true; };
  
  // Create new context
  SwsContext* new_ctx = sws_getContext(
    srcW, srcH, static_cast<AVPixelFormat>(srcFormat),
    dstW, dstH, static_cast<AVPixelFormat>(dstFormat),
    flags, nullptr, nullptr, nullptr
  );
  
  if (!new_ctx) {
    Napi::Error::New(env, "Failed to create SwsContext").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (ctx_ && !is_freed_) { sws_freeContext(ctx_); } ctx_ = new_ctx; is_freed_ = false;
  
  return env.Undefined();
}

Napi::Value SoftwareScaleContext::InitContext(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  SwsContext* ctx = Get();
  if (!ctx) {
    Napi::Error::New(env, "Context not allocated").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  int ret = sws_init_context(ctx, nullptr, nullptr);
  return Napi::Number::New(env, ret);
}

Napi::Value SoftwareScaleContext::FreeContext(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (ctx_ && !is_freed_) { sws_freeContext(ctx_); ctx_ = nullptr; is_freed_ = true; };
  
  return env.Undefined();
}

Napi::Value SoftwareScaleContext::Scale(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  SwsContext* ctx = Get();
  if (!ctx) {
    Napi::Error::New(env, "Context not initialized").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  if (info.Length() < 6) {
    Napi::TypeError::New(env, "Expected 6 arguments (srcSlice, srcStride, srcSliceY, srcSliceH, dst, dstStride)")
        .ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  // Parse source data
  if (!info[0].IsArray() || !info[1].IsArray()) {
    Napi::TypeError::New(env, "srcSlice and srcStride must be arrays").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  Napi::Array srcSlice = info[0].As<Napi::Array>();
  Napi::Array srcStride = info[1].As<Napi::Array>();
  int srcSliceY = info[2].As<Napi::Number>().Int32Value();
  int srcSliceH = info[3].As<Napi::Number>().Int32Value();
  
  // Parse destination data
  if (!info[4].IsArray() || !info[5].IsArray()) {
    Napi::TypeError::New(env, "dst and dstStride must be arrays").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  Napi::Array dst = info[4].As<Napi::Array>();
  Napi::Array dstStride = info[5].As<Napi::Array>();
  
  // Prepare source pointers and strides
  const uint8_t* srcSlicePtr[4] = {nullptr, nullptr, nullptr, nullptr};
  int srcStrideVal[4] = {0, 0, 0, 0};
  
  for (uint32_t i = 0; i < srcSlice.Length() && i < 4; i++) {
    Napi::Value val = srcSlice[i];
    if (val.IsBuffer()) {
      Napi::Buffer<uint8_t> buf = val.As<Napi::Buffer<uint8_t>>();
      srcSlicePtr[i] = buf.Data();
    }
    
    if (i < srcStride.Length()) {
      Napi::Value strideVal = srcStride[i];
      srcStrideVal[i] = strideVal.As<Napi::Number>().Int32Value();
    }
  }
  
  // Prepare destination pointers and strides
  uint8_t* dstPtr[4] = {nullptr, nullptr, nullptr, nullptr};
  int dstStrideVal[4] = {0, 0, 0, 0};
  
  for (uint32_t i = 0; i < dst.Length() && i < 4; i++) {
    Napi::Value val = dst[i];
    if (val.IsBuffer()) {
      Napi::Buffer<uint8_t> buf = val.As<Napi::Buffer<uint8_t>>();
      dstPtr[i] = buf.Data();
    }
    
    if (i < dstStride.Length()) {
      Napi::Value strideVal = dstStride[i];
      dstStrideVal[i] = strideVal.As<Napi::Number>().Int32Value();
    }
  }
  
  // Perform scaling
  int ret = sws_scale(ctx, srcSlicePtr, srcStrideVal, srcSliceY, srcSliceH, dstPtr, dstStrideVal);
  
  return Napi::Number::New(env, ret);
}

Napi::Value SoftwareScaleContext::ScaleFrame(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  SwsContext* ctx = Get();
  if (!ctx) {
    Napi::Error::New(env, "Context not initialized").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  if (info.Length() < 2) {
    Napi::TypeError::New(env, "Expected 2 arguments (dst, src)").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  Frame* dst = UnwrapNativeObject<Frame>(env, info[0], "Frame");
  Frame* src = UnwrapNativeObject<Frame>(env, info[1], "Frame");
  
  if (!dst || !dst->Get() || !src || !src->Get()) {
    Napi::TypeError::New(env, "Invalid frame objects").ThrowAsJavaScriptException();
    return Napi::Number::New(env, AVERROR(EINVAL));
  }
  
  int ret = sws_scale_frame(ctx, dst->Get(), src->Get());
  return Napi::Number::New(env, ret);
}

// === Utility ===

Napi::Value SoftwareScaleContext::Dispose(const Napi::CallbackInfo& info) {
  return FreeContext(info);
}

} // namespace ffmpeg

File: src/lib/software-scale-context.ts
import { SWS_BILINEAR } from '../constants/constants.js';
import { bindings } from './binding.js';
import { OptionMember } from './option.js';

import type { AVPixelFormat, SWSFlag } from '../constants/constants.js';
import type { Frame } from './frame.js';
import type { NativeSoftwareScaleContext, NativeWrapper } from './native-types.js';

/**
 * Video scaling and pixel format conversion context.
 *
 * Provides high-quality image scaling and pixel format conversion for video frames.
 * Supports various scaling algorithms from fast bilinear to high-quality Lanczos.
 * Essential for resolution changes, aspect ratio adjustments, and format compatibility
 * in video processing pipelines.
 *
 * Direct mapping to FFmpeg's SwsContext.
 *
 * @example
 * ```typescript
 * import { SoftwareScaleContext, Frame, FFmpegError } from 'node-av';
 * import { AV_PIX_FMT_YUV420P, AV_PIX_FMT_RGB24, SWS_LANCZOS } from 'node-av/constants';
 *
 * // Create scaler
 * const scaler = new SoftwareScaleContext();
 *
 * // Configure scaling: 1920x1080 YUV420P -> 1280x720 RGB24
 * scaler.getContext(
 *   1920, 1080, AV_PIX_FMT_YUV420P,  // Source
 *   1280, 720, AV_PIX_FMT_RGB24,     // Destination
 *   SWS_LANCZOS                       // High quality
 * );
 *
 * const ret = scaler.initContext();
 * FFmpegError.throwIfError(ret, 'initContext');
 *
 * // Scale frames
 * const dstFrame = new Frame();
 * dstFrame.width = 1280;
 * dstFrame.height = 720;
 * dstFrame.format = AV_PIX_FMT_RGB24;
 * dstFrame.allocBuffer();
 *
 * const height = await scaler.scaleFrame(dstFrame, srcFrame);
 * console.log(`Scaled to ${height} lines`);
 *
 * // Clean up
 * scaler.freeContext();
 * ```
 *
 * @see [SwsContext](https://ffmpeg.org/doxygen/trunk/structSwsContext.html) - FFmpeg Doxygen
 * @see {@link Frame} For video frame operations
 */
export class SoftwareScaleContext extends OptionMember<NativeSoftwareScaleContext> implements Disposable, NativeWrapper<NativeSoftwareScaleContext> {
  constructor() {
    super(new bindings.SoftwareScaleContext());
  }

  /**
   * Allocate scale context.
   *
   * Allocates memory for the scaler.
   * Must be called before configuration if using options.
   *
   * Direct mapping to sws_alloc_context().
   *
   * @example
   * ```typescript
   * const scaler = new SoftwareScaleContext();
   * scaler.allocContext();
   * // Now configure with setOption() or getContext()
   * ```
   *
   * @see {@link getContext} For direct configuration
   */
  allocContext(): void {
    this.native.allocContext();
  }

  /**
   * Configure scaling context.
   *
   * Sets up the scaler with source and destination formats.
   * This is the primary configuration method.
   *
   * Direct mapping to sws_getContext().
   *
   * @param srcW - Source width in pixels
   *
   * @param srcH - Source height in pixels
   *
   * @param srcFormat - Source pixel format
   *
   * @param dstW - Destination width in pixels
   *
   * @param dstH - Destination height in pixels
   *
   * @param dstFormat - Destination pixel format
   *
   * @param flags - Scaling algorithm flags (SWS_*)
   *
   * @example
   * ```typescript
   * import { AV_PIX_FMT_YUV420P, AV_PIX_FMT_RGB24 } from 'node-av/constants';
   * import { SWS_BILINEAR, SWS_BICUBIC, SWS_LANCZOS, SWS_FAST_BILINEAR } from 'node-av/constants';
   *
   * // Fast bilinear (lower quality, faster)
   * scaler.getContext(
   *   1920, 1080, AV_PIX_FMT_YUV420P,
   *   1280, 720, AV_PIX_FMT_RGB24,
   *   SWS_FAST_BILINEAR
   * );
   *
   * // High quality Lanczos (higher quality, slower)
   * scaler.getContext(
   *   1920, 1080, AV_PIX_FMT_YUV420P,
   *   3840, 2160, AV_PIX_FMT_YUV420P,  // Upscaling
   *   SWS_LANCZOS
   * );
   * ```
   *
   * @see {@link initContext} Must be called after configuration
   */
  getContext(srcW: number, srcH: number, srcFormat: AVPixelFormat, dstW: number, dstH: number, dstFormat: AVPixelFormat, flags: SWSFlag = SWS_BILINEAR): void {
    this.native.getContext(srcW, srcH, srcFormat, dstW, dstH, dstFormat, flags);
  }

  /**
   * Initialize scaling context.
   *
   * Initializes the scaler after configuration.
   * Must be called before any scaling operations.
   *
   * Direct mapping to sws_init_context().
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - AVERROR_EINVAL: Invalid parameters
   *   - AVERROR_ENOMEM: Memory allocation failure
   *
   * @example
   * ```typescript
   * import { FFmpegError } from 'node-av';
   *
   * const ret = scaler.initContext();
   * FFmpegError.throwIfError(ret, 'initContext');
   * ```
   *
   * @see {@link getContext} For configuration
   */
  initContext(): number {
    return this.native.initContext();
  }

  /**
   * Free scaling context.
   *
   * Releases all resources associated with the scaler.
   * The context becomes invalid after calling this.
   *
   * Direct mapping to sws_freeContext().
   *
   * @example
   * ```typescript
   * scaler.freeContext();
   * // Scaler is now invalid
   * ```
   *
   * @see {@link Symbol.dispose} For automatic cleanup
   */
  freeContext(): void {
    this.native.freeContext();
  }

  /**
   * Scale image data.
   *
   * Scales raw image data from source to destination buffers.
   * Low-level interface for custom buffer management.
   *
   * Direct mapping to sws_scale().
   *
   * @param srcSlice - Source data planes (one buffer per plane)
   *
   * @param srcStride - Bytes per line for each plane
   *
   * @param srcSliceY - Starting Y position in source
   *
   * @param srcSliceH - Height of source slice to process
   *
   * @param dst - Destination data planes
   *
   * @param dstStride - Destination bytes per line
   *
   * @returns Output height in pixels, negative AVERROR on error:
   *   - AVERROR_EINVAL: Invalid parameters
   *
   * @example
   * ```typescript
   * // Scale YUV420P data
   * const srcPlanes = [yPlane, uPlane, vPlane];
   * const srcStrides = [1920, 960, 960]; // Full HD
   * const dstPlanes = [dstY, dstU, dstV];
   * const dstStrides = [1280, 640, 640]; // 720p
   *
   * const height = await scaler.scale(
   *   srcPlanes, srcStrides, 0, 1080,
   *   dstPlanes, dstStrides
   * );
   * console.log(`Scaled ${height} lines`);
   * ```
   *
   * @see {@link scaleFrame} For frame-based scaling
   */
  async scale(srcSlice: Buffer[], srcStride: number[], srcSliceY: number, srcSliceH: number, dst: Buffer[], dstStride: number[]): Promise<number> {
    return await this.native.scale(srcSlice, srcStride, srcSliceY, srcSliceH, dst, dstStride);
  }

  /**
   * Scale video synchronously.
   * Synchronous version of scale.
   *
   * Scales raw video data from source to destination format.
   * Can scale a slice or entire image.
   *
   * Direct mapping to sws_scale().
   *
   * @param srcSlice - Array of source buffers (one per plane)
   *
   * @param srcStride - Array of source strides (bytes per row)
   *
   * @param srcSliceY - Y position of slice (0 for full image)
   *
   * @param srcSliceH - Height of slice (full height for entire image)
   *
   * @param dst - Array of destination buffers (one per plane)
   *
   * @param dstStride - Array of destination strides
   *
   * @returns Height of output image, or negative AVERROR:
   *   - AVERROR_EINVAL: Invalid parameters
   *
   * @example
   * ```typescript
   * import { FFmpegError } from 'node-av';
   *
   * // Scale YUV420P image
   * const srcBufs = [yPlane, uPlane, vPlane];
   * const srcStrides = [srcWidth, srcWidth/2, srcWidth/2];
   * const dstBufs = [dstY, dstU, dstV];
   * const dstStrides = [dstWidth, dstWidth/2, dstWidth/2];
   *
   * const height = scaler.scaleSync(
   *   srcBufs, srcStrides,
   *   0, srcHeight,  // Full image
   *   dstBufs, dstStrides
   * );
   * FFmpegError.throwIfError(height, 'scaleSync');
   * console.log(`Scaled ${height} lines`);
   * ```
   *
   * @see {@link scale} For async version
   */
  scaleSync(srcSlice: Buffer[], srcStride: number[], srcSliceY: number, srcSliceH: number, dst: Buffer[], dstStride: number[]): number {
    return this.native.scaleSync(srcSlice, srcStride, srcSliceY, srcSliceH, dst, dstStride);
  }

  /**
   * Scale video frame.
   *
   * Scales an entire video frame to the destination format.
   * Simpler interface than scale() for frame-based processing.
   *
   * Direct mapping to sws_scale_frame().
   *
   * @param dst - Destination frame (must be allocated)
   *
   * @param src - Source frame
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - AVERROR_EINVAL: Invalid parameters
   *   - AVERROR_ENOMEM: Memory allocation failure
   *
   * @example
   * ```typescript
   * import { Frame, FFmpegError } from 'node-av';
   * import { AV_PIX_FMT_RGB24 } from 'node-av/constants';
   *
   * // Create destination frame
   * const dstFrame = new Frame();
   * dstFrame.width = 1280;
   * dstFrame.height = 720;
   * dstFrame.format = AV_PIX_FMT_RGB24;
   * const ret = dstFrame.allocBuffer();
   * FFmpegError.throwIfError(ret, 'allocBuffer');
   *
   * // Scale frame
   * const ret2 = await scaler.scaleFrame(dstFrame, srcFrame);
   * FFmpegError.throwIfError(ret2, 'scaleFrame');
   *
   * // dstFrame now contains scaled image
   * ```
   *
   * @see {@link scale} For buffer-based scaling
   */
  async scaleFrame(dst: Frame, src: Frame): Promise<number> {
    return await this.native.scaleFrame(dst.getNative(), src.getNative());
  }

  /**
   * Scale video frame synchronously.
   * Synchronous version of scaleFrame.
   *
   * Scales an entire video frame to the destination format.
   * Simpler interface than scaleSync() for frame-based processing.
   *
   * Direct mapping to sws_scale_frame().
   *
   * @param dst - Destination frame (must be allocated)
   *
   * @param src - Source frame
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - AVERROR_EINVAL: Invalid parameters
   *
   * @example
   * ```typescript
   * import { FFmpegError } from 'node-av';
   * import { AV_PIX_FMT_YUV420P, AV_PIX_FMT_RGB24 } from 'node-av/constants';
   *
   * // Convert YUV to RGB
   * const srcFrame = new Frame();
   * srcFrame.allocBuffer(AV_PIX_FMT_YUV420P, 1920, 1080);
   * // ... fill with YUV data ...
   *
   * const dstFrame = new Frame();
   * dstFrame.allocBuffer(AV_PIX_FMT_RGB24, 1920, 1080);
   *
   * const ret = scaler.scaleFrameSync(dstFrame, srcFrame);
   * FFmpegError.throwIfError(ret, 'scaleFrameSync');
   *
   * // dstFrame now contains scaled image
   * ```
   *
   * @see {@link scaleFrame} For async version
   */
  scaleFrameSync(dst: Frame, src: Frame): number {
    return this.native.scaleFrameSync(dst.getNative(), src.getNative());
  }

  /**
   * Get the underlying native SoftwareScaleContext object.
   *
   * @returns The native SoftwareScaleContext binding object
   *
   * @internal
   */
  getNative(): NativeSoftwareScaleContext {
    return this.native;
  }

  /**
   * Dispose of the scaling context.
   *
   * Implements the Disposable interface for automatic cleanup.
   * Equivalent to calling freeContext().
   *
   * @example
   * ```typescript
   * {
   *   using scaler = new SoftwareScaleContext();
   *   scaler.getContext(...);
   *   scaler.initContext();
   *   // Use scaler...
   * } // Automatically freed when leaving scope
   * ```
   */
  [Symbol.dispose](): void {
    this.native[Symbol.dispose]();
  }
}
